### PR TITLE
Branch on option string names using hash map and switch-case

### PR DESCRIPTION
**Pull request by alex-ozdemir**
We have about 450 options. There are functions that branch on their
string names. Previously, this was done with an if-else chain:

    if (name == "OPTION_1") { return ... }
    if (name == "OPTION_2") { return ... }
    ...

Now, we use a hashmap to enums, followed by a switch-case (jump table):

    OptionEnum t = MAP.at(name);
    switch (t) {
      case OPTION_1: return ...
      case OPTION_2: return ...
      ...
    }

This change seems to reduce the runtime of `get` by about 5x: the unit test that I added previously ran in ~1.23s and now it runs in ~0.24s.

### CODE REVIEW

**Comment by alex-ozdemir** [comment]
@nafur : if you don't want to review this, no problem: we can have someone else do it.

File: src/options/mkoptions.py [hunk 1]
```diff
@@ -305,7 +343,15 @@ def generate_get_impl(modules):
         else:
             ret = '{{ std::stringstream s; s << options.{}.{}; return s.str(); }}'.format(
                 module.id, option.name)
-        res.append('if ({}) {}'.format(cond, ret))
+        res.append("  case OptionEnum::{}:".format(option.enum_name()))
+        res.append("  {")
+        res.append("    {};".format(ret))
+        res.append("  }")
```
Comment (nafur) [inline src/options/mkoptions.py:349]: `ret` is already a single statement (it had no braces in the if)
```suggestion
        res.append("  case OptionEnum::{}: {}".format(option.enum_name(), ret))
```

File: src/options/mkoptions.py [hunk 2]
```diff
@@ -394,8 +449,12 @@ def generate_getinfo_impl(modules):
             constr = 'OptionInfo::ModeInfo{{"{default}", {value}, {{ {modes} }}}}'
         else:
             constr = 'OptionInfo::VoidInfo{{}}'
-        line = 'if ({condition}) return OptionInfo{{"{name}", {{{alias}}}, {setbyuser}, ' + constr + '}};'
+        res.append("  case OptionEnum::{}:".format(option.enum_name()))
+        res.append("  {")
+        line = '    return OptionInfo{{"{name}", {{{alias}}}, {setbyuser}, ' + constr + '}};'
         res.append(line.format(**fmt))
+        res.append("  }")
```
Comment (nafur) [inline src/options/mkoptions.py:456]:
```suggestion
        res.append("  case OptionEnum::{}:".format(option.enum_name()))
        line = '    return OptionInfo{{"{name}", {{{alias}}}, {setbyuser}, ' + constr + '}};'
        res.append(line.format(**fmt))
```

**Review by nafur — APPROVED** [review]
I like this change!
My comments are really nit-picky, ignore them if you want.
Out of curiosity:
- have you checked the memory footprint of `NAME_TO_ENUM`?
- have you benchmarked with `std::map` instead `std::unordered_map`?

File: src/options/mkoptions.py [hunk 3]
```diff
@@ -272,11 +275,42 @@ def generate_holder_mem_copy(modules):
...
+
+    When we measured, this was about 5x faster than a huge if-else chain.
+    It would probably be even faster with a better hash function.
+    """
+    res = []
+    res.append("enum class OptionEnum {")
+    for module, option in all_options(modules, True):
+        if not option.long:
+            continue
+        res.append("  {n},".format(n=option.enum_name()))
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply src/options/mkoptions.py]: I changed the quotes, but didn't use f-strings. Best to change the whole file at once.

**Pull request merged**